### PR TITLE
Support clustering agents

### PR DESCRIPTION
**Pull request by rfratto**
This PR completes the work from #42 and introduces the remainder of the scraping service mode. Agents will cluster using the Cortex Ring and watch the config API for changes. Tokens from the Cortex Ring will be used to figure out which agents are responsible for which configs. Such configs will be dynamically added to the agent.

The current implementation doesn't read from the store on startup; this will be fixed before the PR is moved out of draft. For now, if testing locally, keep in mind that you should only apply new configs after all nodes are up in the cluster. 

This PR shuffles around a bunch of the code in `pkg/prometheus`. It is now split into four main packages:

- `pkg/prometheus`: Runs a collection of instances
- `pkg/prometheus/ha`: Scraping service code and instance config API. Notifies `pkg/prometheus` when a new instance should be started, stopped, or restarted with a new config 
- `pkg/prometheus/instance`: Code for an individual instance (i.e., combination of Prometheus SD, WAL, remote_write from one config)
- `pkg/prometheus/wal`: WAL appender

TODO: 

- [x] Watch KV store for changes, spawn instances based on added/removed configs
- [x] Added docker-compose config to `example/` to play around with the mode.
- [x] Cluster reshard on join/leave
- [x] Automated tests 
- [x] Tooling in `example/` to deploy all instance configs
- [x] Documentation 

MANUAL TESTING: 

- [x] Verify that Agents respond to configs added to the store 
- [x] Verify that Agents respond to configs changed in the store 
- [x] Verify that Agents respond to configs removed from the store
- [x] Verify that Agents reshard on scale up/scale down
- [x] Verify that Agents pick up existing configs in store after starting up

### CODE REVIEW

File: pkg/prometheus/ha/server.go [hunk 1]
```diff
@@ -259,11 +237,7 @@ func (s *Server) notifyReshard(ctx context.Context, desc *ring.IngesterDesc) err
 	}
 	defer cli.Close()
 
-	backoff := util.NewBackoff(context.Background(), util.BackoffConfig{
-		MinBackoff: time.Second,
-		MaxBackoff: 2 * time.Minute,
-		MaxRetries: 10,
-	})
+	backoff := util.NewBackoff(context.Background(), backoffConfig)
```
Comment (hoenn) [inline pkg/prometheus/ha/server.go:240]: What's the reason for passing `context.Background()` here instead of the original `ctx`?

Comment (rfratto) [reply pkg/prometheus/ha/server.go:240]: I was blindly copying things :) Nice catch, I'll fix it.

File: pkg/prometheus/agent.go [hunk 2]
```diff
@@ -117,7 +108,7 @@ type Agent struct {
 
 	instanceFactory instanceFactory
 
-	kv kv.Client
+	ha *ha.Server
```
Comment (joe-elliott) [inline pkg/prometheus/agent.go:111]: why is this so funny?  i don't think we should be so flippant about our variables.

**Comment by rfratto** [comment]
This should be changed to always set the replication factor to 1 so it doesn't have to be set in the config.

**Comment by rfratto** [comment]
I'm going to mark this as ready for review, though I still have to write documentation and new dashboards.

**Comment by rfratto** [comment]
@joe-elliott @hoenn If one of you are able and willing, I'd appreciate a review of the [scraping service docs](https://github.com/rfratto/agent/blob/cluster-agents/docs/scraping-service.md). 

I'm not trying to make it perfect yet; I plan to have an operators guide later, but I want a doc that:

1) explains what the scraping service is 
2) explains how the scraping service works (esp around the hash ring)

Let me know if you think I've been able to do that here or if I can add any new info.

File: docs/scraping-service.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
+# Scraping Service Mode
+
+Scraping Service Mode is a third operational mode of the Grafana Cloud Agent
+that allows for clustering a set of Agent processes and distributing scrape load
+between them.
```
Comment (cstyan) [inline docs/scraping-service.md:5]: 'across' might be a better word to use here

File: docs/scraping-service.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+Determining what to scrape is done by storing instance configuration files to an
+[API](./api.md) which then stores the configuration files in a KV store backend.
+All agents in the cluster must use the same KV store so they read the same set
```
Comment (cstyan) [inline docs/scraping-service.md:9]: maybe bold must?

File: docs/scraping-service.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+of config files.
+
+The "instance" refers to the process of:
+
+- Service discovery for all `scrape_configs` within that loaded config
+- Scrapes metrics from all discovered targets
+- Stores data in its own Write-Ahead Log specific to the loaded config
+- Remote Writes scraped metrics to the configured `remote_write` clients
+  specified within the loaded config.
+
```
Comment (cstyan) [inline docs/scraping-service.md:19]: Maybe add one final line after the bullet points `This means that, for each unique config file an agent is supposed to collect metrics for, there is an "instance" running each of these above processes`

Comment (cstyan) [reply docs/scraping-service.md:19]: Oh okay, you clarify that in the next sentence. So maybe just consider reordering here, but I don't have a concrete suggestion for that.

File: docs/scraping-service.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+The "instance configuration file," then, is the configuration file that
```
Comment (cstyan) [inline docs/scraping-service.md:21]: ah short example of what (or why you would use) multiple "instance" configs might look like may be useful

File: docs/scraping-service.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+specifies the set of `scrape_configs` and `remote_write` endpoints. Each process
+of the Grafana Cloud Agent can be running multiple independent instances at
+once.
+
+When Scraping Service Mode is enabled, Agents disallow specifying
```
Comment (cstyan) [inline docs/scraping-service.md:26]: again maybe consider bolding that local config is disallowed?

File: docs/scraping-service.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+sharding. The default number of generated tokens is 128.
+
+The Distributed Hash Ring is also stored in a KV store. Since a KV store is
+also needed for storing configuration files, it is common practice to re-use
+that same KV store for the ring.
+
+When sharding, the Agent currently uses the entire contents of a config file
+stored in the KV store for load distribution. The hash of the config file is
+used as the _key_ in the ring and looks up an Agent that should be responsible
+for that config. "Price is Right" rules are used for the Agent lookup; the Agent
```
Comment (cstyan) [inline docs/scraping-service.md:47]: This "looks up an Agent that should be responsible", it's not clear who/which process is doing the lookup here

**Comment by rfratto** [comment]
@cstyan thanks for reviewing! I addressed all your feedback, PTAL

**Pull request merged**